titan-compiler/ast.lua

```diff
@@ -54,21 +54,32 @@ types.Args = {
 
```

Comment:
> inspect output might be less readable.

This is a major con. Not only for inspect, but for the rest of the code that uses the AST. See the Typed Lua codebase, full of stuff like `stat[1][3][2][1].type == stat[1][3][1].type`. It takes someone deeply entrenched in that codebase like @andremm was during his PhD to be able to glance that and understand what does it mean (and I doubt he still remembers the meaning of all those numbers there as quickly today!).

Reply:
> See the Typed Lua codebase, full of stuff like stat[1][3][2][1].type == stat[1][3][1].type

@hishamhm I'm not sure where did you find this kind of construction. I just ran `grep -r "stm\[" * | grep "\.type"`  in the Typed Lua code base and couldn't find them. :)
I used `grep` with `stm` instead of `stat` because it is `stm` that I use to access statement nodes instead of using `stat`.

Anyway, I agree that inspect output being less readable is a major con.

Reply:
@andremm I was paraphrasing :) And the code has changed a lot since I last played with it! I had to dig back in to the Git history, I was thinking of stuff like this: https://github.com/andremm/typedlua/blob/9b9bd5d1ba4f8069830302fefb7ff91cf3ba2343/typedlua/tlchecker.lua#L1156-L1160 (this is the most pathological example I ran into, but there's lots of stuff like `[2][1]` out there which assume you have the AST format in your head.)

Reply:
hahahaha, yeah, you have a very good point @hishamhm. The first versions of the code that implemented the filters of union types inside if statements were kind of convoluted and assumed you had the AST stuck in your mind. :)